# Risk Assessment and Creditworthiness App



Import the necessary Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, accuracy_score, precision_score, classification_report, recall_score, f1_score, roc_auc_score, auc


To prevent the warning error from showing

In [3]:
import warnings
warnings.filterwarnings('ignore')

# Explanation of the Dataset

The definition of the variables are as follows;

person_age: The age of the borrower when securing the loan.

person_income: The borrower’s annual earnings at the time of the loan.

person_home_ownership: Type of home ownership of the applicant;
Rent: The borrower is currently renting a property.
Mortgage: The borrower has a mortgage on the property they own.
Own: The borrower owns a home outright.
Other: Other categories of home ownership that may be specific to the dataset.

person_emp_length: The amount of time in years that the borrower has been in employment.

loan_intent: Loan purpose - what applicant (borrower) wants to use the loan for.

loan_grade: A classification system based on credit history, collateral quality, and the likelihood of repayment of the principal and interest.
A: The borrower has a high creditworthiness, indicating low risk.
B: The borrower is relatively low-risk, but not as creditworthy as Grade A.
C: The borrower’s creditworthiness is moderate.
D: The borrower is considered to have a higher risk compared to previous grades.
E: The borrower’s creditworthiness is lower, indicating a higher risk.
F: The borrower poses a significant credit risk.
G: The borrower’s creditworthiness is the lowest, signifying the highest risk.

loan_amnt: Total amount of the loan applied for.

loan_int_rate: The interest rate of the loan.

loan_status: Target variable indicating Default as (1) or Non-default as (0). The loan_status variable is a crucial dependent variable. A default occurs when a borrower is unable to make timely payments, misses payments, or avoids or stops making payments on interest or principal owed;
0: Non-default - The borrower successfully repaid the loan as agreed, and there was no default.
1: Default - The borrower failed to repay the loan according to the agreed-upon terms and defaulted.

loan_percent_income: Ratio between the loan amount and the annual income of borrower.

cb_person_cred_hist_length: The number of years of personal history since the first loan was taken by the borrower.

cb_person_default_on_file: Indicates if the person has previously defaulted.


# The DataFrame is referred to as "dfrisk"

In [4]:
dfrisk = pd.read_csv('C:/Users/PC/Documents/sola/others/Telegram Desktop/credit_risk_dataset.csv')
dfrisk

person_age  person_income person_home_ownership  person_emp_length  \
0              22          59000                  RENT              123.0   
1              21           9600                   OWN                5.0   
2              25           9600              MORTGAGE                1.0   
3              23          65500                  RENT                4.0   
4              24          54400                  RENT                8.0   
...           ...            ...                   ...                ...   
32576          57          53000              MORTGAGE                1.0   
32577          54         120000              MORTGAGE                4.0   
32578          65          76000                  RENT                3.0   
32579          56         150000              MORTGAGE                5.0   
32580          66          42000                  RENT                2.0   

           loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0             PERSONAL          D      35000          16.02            1   
1            EDUCATION          B       1000          11.14            0   
2              MEDICAL          C       5500          12.87            1   
3              MEDICAL          C      35000          15.23            1   
4              MEDICAL          C      35000          14.27            1   
...                ...        ...        ...            ...          ...   
32576         PERSONAL          C       5800          13.16            0   
32577         PERSONAL          A      17625           7.49            0   
32578  HOMEIMPROVEMENT          B      35000          10.99            1   
32579         PERSONAL          B      15000          11.48            0   
32580          MEDICAL          B       6475           9.99            0   

       loan_percent_income cb_person_default_on_file  \
0                     0.59                         Y   
1                     0.10                         N   
2                     0.57                         N   
3                     0.53                         N   
4                     0.55                         Y   
...                    ...                       ...   
32576                 0.11                         N   
32577                 0.15                         N   
32578                 0.46                         N   
32579                 0.10                         N   
32580                 0.15                         N   

       cb_person_cred_hist_length  
0                               3  
1                               2  
2                               3  
3                               2  
4                               4  
...                           ...  
32576                          30  
32577                          19  
32578                          28  
32579                          26  
32580                          30  

[32581 rows x 12 columns]

In [5]:
dfrisk.describe()

person_age  person_income  person_emp_length     loan_amnt  \
count  32581.000000   3.258100e+04       31686.000000  32581.000000   
mean      27.734600   6.607485e+04           4.789686   9589.371106   
std        6.348078   6.198312e+04           4.142630   6322.086646   
min       20.000000   4.000000e+03           0.000000    500.000000   
25%       23.000000   3.850000e+04           2.000000   5000.000000   
50%       26.000000   5.500000e+04           4.000000   8000.000000   
75%       30.000000   7.920000e+04           7.000000  12200.000000   
max      144.000000   6.000000e+06         123.000000  35000.000000   

       loan_int_rate   loan_status  loan_percent_income  \
count   29465.000000  32581.000000         32581.000000   
mean       11.011695      0.218164             0.170203   
std         3.240459      0.413006             0.106782   
min         5.420000      0.000000             0.000000   
25%         7.900000      0.000000             0.090000   
50%        10.990000      0.000000             0.150000   
75%        13.470000      0.000000             0.230000   
max        23.220000      1.000000             0.830000   

       cb_person_cred_hist_length  
count                32581.000000  
mean                     5.804211  
std                      4.055001  
min                      2.000000  
25%                      3.000000  
50%                      4.000000  
75%                      8.000000  
max                     30.000000

Capture the data types of each column

In [6]:
dfrisk.dtypes

person_age                      int64
person_income                   int64
person_home_ownership          object
person_emp_length             float64
loan_intent                    object
loan_grade                     object
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file      object
cb_person_cred_hist_length      int64
dtype: object

# Preprocessing

Instead of making use of the encoder, we manually converted the categorical columns into Numerical columns

categorical_columns ==>> 'person_home_ownership', 'loan_intent', 'loan_grade', 'cb_person_default_on_file'

Column >> 'person_home_ownership'

In [7]:
dfrisk['person_home_ownership'].replace('RENT',0,inplace=True)
dfrisk['person_home_ownership'].replace('OWN',1,inplace=True)
dfrisk['person_home_ownership'].replace('MORTGAGE',2,inplace=True)
dfrisk['person_home_ownership'].replace('OTHER',3,inplace=True)

In [8]:
dfrisk['person_home_ownership'].value_counts()

person_home_ownership
0    16446
2    13444
1     2584
3      107
Name: count, dtype: int64

Column >> 'loan_intent'

In [9]:
dfrisk['loan_intent'].replace('PERSONAL',0,inplace=True)
dfrisk['loan_intent'].replace('EDUCATION',1,inplace=True)
dfrisk['loan_intent'].replace('MEDICAL',2,inplace=True)
dfrisk['loan_intent'].replace('VENTURE',3,inplace=True)
dfrisk['loan_intent'].replace('HOMEIMPROVEMENT',4,inplace=True)
dfrisk['loan_intent'].replace('DEBTCONSOLIDATION',5,inplace=True)

In [10]:
dfrisk['loan_intent'].value_counts()

loan_intent
1    6453
2    6071
3    5719
0    5521
5    5212
4    3605
Name: count, dtype: int64

Column >> 'loan_grade'

In [11]:
dfrisk['loan_grade'].replace('A',0,inplace=True)
dfrisk['loan_grade'].replace('B',1,inplace=True)
dfrisk['loan_grade'].replace('C',2,inplace=True)
dfrisk['loan_grade'].replace('D',3,inplace=True)
dfrisk['loan_grade'].replace('E',4,inplace=True)
dfrisk['loan_grade'].replace('F',5,inplace=True)
dfrisk['loan_grade'].replace('G',6,inplace=True)

In [12]:
dfrisk['loan_grade'].value_counts()

loan_grade
0    10777
1    10451
2     6458
3     3626
4      964
5      241
6       64
Name: count, dtype: int64

Column >> 'cb_person_default_on_file'

In [13]:
dfrisk['cb_person_default_on_file'].replace('Y',1,inplace=True)
dfrisk['cb_person_default_on_file'].replace('N',0,inplace=True)

In [14]:
dfrisk['cb_person_default_on_file'].value_counts()

cb_person_default_on_file
0    26836
1     5745
Name: count, dtype: int64

Generate the unique values of the converted categorical columns to confirm correctness

In [15]:
dfrisk['loan_intent'].unique()

array([0, 1, 2, 3, 4, 5], dtype=int64)

In [16]:
dfrisk['loan_grade'].unique()

array([3, 1, 2, 0, 4, 5, 6], dtype=int64)

In [17]:
dfrisk['cb_person_default_on_file'].unique()

array([1, 0], dtype=int64)

In [18]:
dfrisk['person_home_ownership'].unique()

array([0, 1, 2, 3], dtype=int64)

Another glance at dataset, in order to further Preprocess

In [19]:
dfrisk.describe()

person_age  person_income  person_home_ownership  person_emp_length  \
count  32581.000000   3.258100e+04           32581.000000       31686.000000   
mean      27.734600   6.607485e+04               0.914429           4.789686   
std        6.348078   6.198312e+04               0.960858           4.142630   
min       20.000000   4.000000e+03               0.000000           0.000000   
25%       23.000000   3.850000e+04               0.000000           2.000000   
50%       26.000000   5.500000e+04               0.000000           4.000000   
75%       30.000000   7.920000e+04               2.000000           7.000000   
max      144.000000   6.000000e+06               3.000000         123.000000   

        loan_intent    loan_grade     loan_amnt  loan_int_rate   loan_status  \
count  32581.000000  32581.000000  32581.000000   29465.000000  32581.000000   
mean       2.339769      1.218195   9589.371106      11.011695      0.218164   
std        1.678803      1.166336   6322.086646       3.240459      0.413006   
min        0.000000      0.000000    500.000000       5.420000      0.000000   
25%        1.000000      0.000000   5000.000000       7.900000      0.000000   
50%        2.000000      1.000000   8000.000000      10.990000      0.000000   
75%        4.000000      2.000000  12200.000000      13.470000      0.000000   
max        5.000000      6.000000  35000.000000      23.220000      1.000000   

       loan_percent_income  cb_person_default_on_file  \
count         32581.000000               32581.000000   
mean              0.170203                   0.176330   
std               0.106782                   0.381106   
min               0.000000                   0.000000   
25%               0.090000                   0.000000   
50%               0.150000                   0.000000   
75%               0.230000                   0.000000   
max               0.830000                   1.000000   

       cb_person_cred_hist_length  
count                32581.000000  
mean                     5.804211  
std                      4.055001  
min                      2.000000  
25%                      3.000000  
50%                      4.000000  
75%                      8.000000  
max                     30.000000

Confirmation that all 4 categorical columns have been transformed into numerical...

In [20]:
print(dfrisk.dtypes)

person_age                      int64
person_income                   int64
person_home_ownership           int64
person_emp_length             float64
loan_intent                     int64
loan_grade                      int64
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file       int64
cb_person_cred_hist_length      int64
dtype: object


Checking the number of NaN values per column

In [21]:
dfrisk.isna().sum()

person_age                       0
person_income                    0
person_home_ownership            0
person_emp_length              895
loan_intent                      0
loan_grade                       0
loan_amnt                        0
loan_int_rate                 3116
loan_status                      0
loan_percent_income              0
cb_person_default_on_file        0
cb_person_cred_hist_length       0
dtype: int64

Replace the NaN values in the 'person_emp_length' column with the median value

In [22]:
dfrisk['person_emp_length'].fillna(dfrisk['person_emp_length'].median(), inplace=True)

Confirm that the 'person_emp_length' column has no NaN values anymore

In [23]:
dfrisk.isna().sum()

person_age                       0
person_income                    0
person_home_ownership            0
person_emp_length                0
loan_intent                      0
loan_grade                       0
loan_amnt                        0
loan_int_rate                 3116
loan_status                      0
loan_percent_income              0
cb_person_default_on_file        0
cb_person_cred_hist_length       0
dtype: int64

Replace the NaN values in the 'loan_int_rate' column with the mode value

In [24]:
dfrisk['loan_int_rate'].fillna(dfrisk['loan_int_rate'].median(), inplace=True)

Confirm that the 'loan_int_rate' column has no NaN values anymore

In [25]:
dfrisk.isna().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

Remove unrealistic fields from 2 identified columns - 144 years of age and 123 years of service

In [26]:
#Check Modal age in person_age by calculating the mode of 'person_age' column
mode_person_age = dfrisk['person_age'].mode()

# Update the 'person_age' column where the age is greater than 75 with the mode age
dfrisk.loc[dfrisk['person_age'] > 75, 'person_age'] = mode_person_age

In [27]:
# Calculate the Median value for 'person_emp_length' column
median_person_emp_length = dfrisk['person_emp_length'].median()

# Update the 'person_emp_length' column where it is greater than 50 with the median person_emp_length
dfrisk.loc[dfrisk['person_emp_length'] > 50, 'person_emp_length'] = median_person_emp_length

Confirm the removal of outliers, another look at the dataset

In [28]:
dfrisk

person_age  person_income  person_home_ownership  person_emp_length  \
0            22.0          59000                      0                4.0   
1            21.0           9600                      1                5.0   
2            25.0           9600                      2                1.0   
3            23.0          65500                      0                4.0   
4            24.0          54400                      0                8.0   
...           ...            ...                    ...                ...   
32576        57.0          53000                      2                1.0   
32577        54.0         120000                      2                4.0   
32578        65.0          76000                      0                3.0   
32579        56.0         150000                      2                5.0   
32580        66.0          42000                      0                2.0   

       loan_intent  loan_grade  loan_amnt  loan_int_rate  loan_status  \
0                0           3      35000          16.02            1   
1                1           1       1000          11.14            0   
2                2           2       5500          12.87            1   
3                2           2      35000          15.23            1   
4                2           2      35000          14.27            1   
...            ...         ...        ...            ...          ...   
32576            0           2       5800          13.16            0   
32577            0           0      17625           7.49            0   
32578            4           1      35000          10.99            1   
32579            0           1      15000          11.48            0   
32580            2           1       6475           9.99            0   

       loan_percent_income  cb_person_default_on_file  \
0                     0.59                          1   
1                     0.10                          0   
2                     0.57                          0   
3                     0.53                          0   
4                     0.55                          1   
...                    ...                        ...   
32576                 0.11                          0   
32577                 0.15                          0   
32578                 0.46                          0   
32579                 0.10                          0   
32580                 0.15                          0   

       cb_person_cred_hist_length  
0                               3  
1                               2  
2                               3  
3                               2  
4                               4  
...                           ...  
32576                          30  
32577                          19  
32578                          28  
32579                          26  
32580                          30  

[32581 rows x 12 columns]

In [29]:
dfrisk.head()

person_age  person_income  person_home_ownership  person_emp_length  \
0        22.0          59000                      0                4.0   
1        21.0           9600                      1                5.0   
2        25.0           9600                      2                1.0   
3        23.0          65500                      0                4.0   
4        24.0          54400                      0                8.0   

   loan_intent  loan_grade  loan_amnt  loan_int_rate  loan_status  \
0            0           3      35000          16.02            1   
1            1           1       1000          11.14            0   
2            2           2       5500          12.87            1   
3            2           2      35000          15.23            1   
4            2           2      35000          14.27            1   

   loan_percent_income  cb_person_default_on_file  cb_person_cred_hist_length  
0                 0.59                          1                           3  
1                 0.10                          0                           2  
2                 0.57                          0                           3  
3                 0.53                          0                           2  
4                 0.55                          1                           4

In [30]:
dfrisk.tail()

person_age  person_income  person_home_ownership  person_emp_length  \
32576        57.0          53000                      2                1.0   
32577        54.0         120000                      2                4.0   
32578        65.0          76000                      0                3.0   
32579        56.0         150000                      2                5.0   
32580        66.0          42000                      0                2.0   

       loan_intent  loan_grade  loan_amnt  loan_int_rate  loan_status  \
32576            0           2       5800          13.16            0   
32577            0           0      17625           7.49            0   
32578            4           1      35000          10.99            1   
32579            0           1      15000          11.48            0   
32580            2           1       6475           9.99            0   

       loan_percent_income  cb_person_default_on_file  \
32576                 0.11                          0   
32577                 0.15                          0   
32578                 0.46                          0   
32579                 0.10                          0   
32580                 0.15                          0   

       cb_person_cred_hist_length  
32576                          30  
32577                          19  
32578                          28  
32579                          26  
32580                          30

In [31]:
dfrisk.describe()

person_age  person_income  person_home_ownership  person_emp_length  \
count  32571.000000   3.258100e+04           32581.000000       32581.000000   
mean      27.709650   6.607485e+04               0.914429           4.760689   
std        6.167858   6.198312e+04               0.960858           3.981013   
min       20.000000   4.000000e+03               0.000000           0.000000   
25%       23.000000   3.850000e+04               0.000000           2.000000   
50%       26.000000   5.500000e+04               0.000000           4.000000   
75%       30.000000   7.920000e+04               2.000000           7.000000   
max       73.000000   6.000000e+06               3.000000          41.000000   

        loan_intent    loan_grade     loan_amnt  loan_int_rate   loan_status  \
count  32581.000000  32581.000000  32581.000000   32581.000000  32581.000000   
mean       2.339769      1.218195   9589.371106      11.009620      0.218164   
std        1.678803      1.166336   6322.086646       3.081611      0.413006   
min        0.000000      0.000000    500.000000       5.420000      0.000000   
25%        1.000000      0.000000   5000.000000       8.490000      0.000000   
50%        2.000000      1.000000   8000.000000      10.990000      0.000000   
75%        4.000000      2.000000  12200.000000      13.110000      0.000000   
max        5.000000      6.000000  35000.000000      23.220000      1.000000   

       loan_percent_income  cb_person_default_on_file  \
count         32581.000000               32581.000000   
mean              0.170203                   0.176330   
std               0.106782                   0.381106   
min               0.000000                   0.000000   
25%               0.090000                   0.000000   
50%               0.150000                   0.000000   
75%               0.230000                   0.000000   
max               0.830000                   1.000000   

       cb_person_cred_hist_length  
count                32581.000000  
mean                     5.804211  
std                      4.055001  
min                      2.000000  
25%                      3.000000  
50%                      4.000000  
75%                      8.000000  
max                     30.000000

In [32]:
dfrisk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32571 non-null  float64
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  int64  
 3   person_emp_length           32581 non-null  float64
 4   loan_intent                 32581 non-null  int64  
 5   loan_grade                  32581 non-null  int64  
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               32581 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  int64  
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(4), int64(8)
memory usage: 3.0 MB


In [33]:
print(dfrisk)

       person_age  person_income  person_home_ownership  person_emp_length  \
0            22.0          59000                      0                4.0   
1            21.0           9600                      1                5.0   
2            25.0           9600                      2                1.0   
3            23.0          65500                      0                4.0   
4            24.0          54400                      0                8.0   
...           ...            ...                    ...                ...   
32576        57.0          53000                      2                1.0   
32577        54.0         120000                      2                4.0   
32578        65.0          76000                      0                3.0   
32579        56.0         150000                      2                5.0   
32580        66.0          42000                      0                2.0   

       loan_intent  loan_grade  loan_amnt  loan_int_rate  loan_

# LogisticRegression

Import necessary Libraries

In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier, XGBRegressor
from sklearn.metrics import accuracy_score, mean_squared_error, mean_absolute_error, r2_score

Check data types

In [35]:
print(dfrisk.dtypes)

person_age                    float64
person_income                   int64
person_home_ownership           int64
person_emp_length             float64
loan_intent                     int64
loan_grade                      int64
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file       int64
cb_person_cred_hist_length      int64
dtype: object


Prepare features and target for Logistic Regression

In [36]:
dfrisk.corr()

person_age  person_income  person_home_ownership  \
person_age                    1.000000       0.141424               0.033113   
person_income                 0.141424       1.000000               0.199315   
person_home_ownership         0.033113       0.199315               1.000000   
person_emp_length             0.168988       0.137016               0.233845   
loan_intent                   0.027972       0.016113               0.017682   
loan_grade                    0.014779      -0.001022              -0.119541   
loan_amnt                     0.052195       0.266820               0.128938   
loan_int_rate                 0.012065       0.000746              -0.132323   
loan_status                  -0.020128      -0.144449              -0.218714   
loan_percent_income          -0.041733      -0.254471              -0.135376   
cb_person_default_on_file     0.007850      -0.003613              -0.061556   
cb_person_cred_hist_length    0.878456       0.117987               0.025155   

                            person_emp_length  loan_intent  loan_grade  \
person_age                           0.168988     0.027972    0.014779   
person_income                        0.137016     0.016113   -0.001022   
person_home_ownership                0.233845     0.017682   -0.119541   
person_emp_length                    1.000000     0.017365   -0.046959   
loan_intent                          0.017365     1.000000    0.013898   
loan_grade                          -0.046959     0.013898    1.000000   
loan_amnt                            0.111694     0.018725    0.145799   
loan_int_rate                       -0.052793     0.006006    0.889929   
loan_status                         -0.085630     0.065381    0.373080   
loan_percent_income                 -0.058584    -0.001939    0.123021   
cb_person_default_on_file           -0.029034     0.010508    0.537054   
cb_person_cred_hist_length           0.147934     0.016905    0.015069   

                            loan_amnt  loan_int_rate  loan_status  \
person_age                   0.052195       0.012065    -0.020128   
person_income                0.266820       0.000746    -0.144449   
person_home_ownership        0.128938      -0.132323    -0.218714   
person_emp_length            0.111694      -0.052793    -0.085630   
loan_intent                  0.018725       0.006006     0.065381   
loan_grade                   0.145799       0.889929     0.373080   
loan_amnt                    1.000000       0.139483     0.105376   
loan_int_rate                0.139483       1.000000     0.319360   
loan_status                  0.105376       0.319360     1.000000   
loan_percent_income          0.572612       0.114514     0.379366   
cb_person_default_on_file    0.039081       0.477146     0.179141   
cb_person_cred_hist_length   0.041967       0.015762    -0.015529   

                            loan_percent_income  cb_person_default_on_file  \
person_age                            -0.041733                   0.007850   
person_income                         -0.254471                  -0.003613   
person_home_ownership                 -0.135376                  -0.061556   
person_emp_length                     -0.058584                  -0.029034   
loan_intent                           -0.001939                   0.010508   
loan_grade                             0.123021                   0.537054   
loan_amnt                              0.572612                   0.039081   
loan_int_rate                          0.114514                   0.477146   
loan_status                            0.379366                   0.179141   
loan_percent_income                    1.000000                   0.035910   
cb_person_default_on_file              0.035910                   1.000000   
cb_person_cred_hist_length            -0.031690                   0.004564   

                            cb_person_cred_hist_length  
person_age                                    0.878456

In [37]:
dfrisk.isna().sum()

person_age                    10
person_income                  0
person_home_ownership          0
person_emp_length              0
loan_intent                    0
loan_grade                     0
loan_amnt                      0
loan_int_rate                  0
loan_status                    0
loan_percent_income            0
cb_person_default_on_file      0
cb_person_cred_hist_length     0
dtype: int64

Fill the NaN values with the Median Value of the person_age column

In [38]:
dfrisk['person_age'].fillna(dfrisk['person_age'].median(), inplace=True)

In [39]:
dfrisk.isna().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

In [40]:
X = dfrisk.drop(['loan_status'], axis=1)
y = dfrisk['loan_status']

Split the dataset >> dfrisk

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Perform standardization

In [42]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_train_scaled[:5]

array([[ 5.8452654 , -0.31717677, -0.9521391 , -0.69239938, -1.39203199,
         0.6746644 , -0.76018887,  0.02663141, -0.65843875,  2.17447733,
         4.48497655],
       [-0.28026618, -0.63165151,  0.08886738, -1.19395427,  1.57848724,
         2.39492002, -0.17434095,  1.77173414,  1.49131123, -0.45988063,
        -0.69475502],
       [-0.76386077, -0.23855808,  1.12987386, -0.44162193, -1.39203199,
         0.6746644 ,  1.01318862,  0.68430072,  1.30437645,  2.17447733,
        -0.69475502],
       [-0.92505897, -0.1598765 ,  1.12987386,  0.31071041, -0.2038243 ,
        -1.04559122, -0.57018414, -1.01847675, -0.56497136, -0.45988063,
        -0.44810114],
       [-0.60266257,  0.20170655, -0.9521391 , -0.44162193, -1.39203199,
         0.6746644 , -0.41184686,  0.49872077, -0.75190614, -0.45988063,
        -0.69475502]])

Instantiate the LogisticRegression model

In [43]:
model = LogisticRegression(max_iter=1000)

Fit the model

In [44]:
model.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

Predict the model on the test set

In [45]:
y_pred = model.predict(X_test_scaled)

In [46]:
y_pred

array([1, 0, 0, ..., 0, 0, 0], dtype=int64)

In [47]:
y_test

14668    0
24614    0
11096    0
10424    1
26007    1
        ..
31330    0
2862     0
14754    0
14170    0
24385    0
Name: loan_status, Length: 6517, dtype: int64

Evaluate model performance

In [48]:

accuracy = accuracy_score(y_test, y_pred)
print(f"The LogisticRegression Model performance Accuracy => {round(accuracy, 2)}")

The LogisticRegression Model performance Accuracy => 0.84


Let's take a look at other Evaluation Metrics

In [49]:
logisticRegression_report = classification_report(y_test, y_pred)
print("Classification Report for LogisticRegression \n")
print(f"{logisticRegression_report}")

Classification Report for LogisticRegression 

              precision    recall  f1-score   support

           0       0.86      0.95      0.90      5072
           1       0.72      0.47      0.56      1445

    accuracy                           0.84      6517
   macro avg       0.79      0.71      0.73      6517
weighted avg       0.83      0.84      0.83      6517



In [50]:
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r_square = r2_score(y_test, y_pred)

In [51]:
print(f"The LogisticRegression Model MSE => {round(mse, 2)}")
print(f"The LogisticRegression Model MAE => {round(mae, 2)}")
print(f"The LogisticRegression Model R-Square => {round(r_square, 2)}")

The LogisticRegression Model MSE => 0.16
The LogisticRegression Model MAE => 0.16
The LogisticRegression Model R-Square => 0.08


Save the LogisticRegression model

In [52]:
import joblib


joblib.dump(model, 'C:/Users/PC/Documents/sola/others/Telegram Desktop/logistic_regression_model.pkl')

['C:/Users/PC/Documents/sola/others/Telegram Desktop/logistic_regression_model.pkl']

# - - - End of LogisticRegression - - -

# XGBoost Regression 'XRegressor'

Define features and target

In [53]:

X = dfrisk.drop(columns=['loan_amnt'])
y = dfrisk['loan_amnt']

Split the dataset >> dfrisk

In [54]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Perform standardization

In [55]:

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_train_scaled[:5]

array([[ 5.8452654 , -0.31717677, -0.9521391 , -0.69239938, -1.39203199,
         0.6746644 ,  0.02663141, -0.52686282, -0.65843875,  2.17447733,
         4.48497655],
       [-0.28026618, -0.63165151,  0.08886738, -1.19395427,  1.57848724,
         2.39492002,  1.77173414,  1.89802727,  1.49131123, -0.45988063,
        -0.69475502],
       [-0.76386077, -0.23855808,  1.12987386, -0.44162193, -1.39203199,
         0.6746644 ,  0.68430072, -0.52686282,  1.30437645,  2.17447733,
        -0.69475502],
       [-0.92505897, -0.1598765 ,  1.12987386,  0.31071041, -0.2038243 ,
        -1.04559122, -1.01847675, -0.52686282, -0.56497136, -0.45988063,
        -0.44810114],
       [-0.60266257,  0.20170655, -0.9521391 , -0.44162193, -1.39203199,
         0.6746644 ,  0.49872077, -0.52686282, -0.75190614, -0.45988063,
        -0.69475502]])

Import XGBoost Algorithm

In [56]:
import xgboost as xgb

Initialize and train the XGBoost regressor model


In [57]:
xgb_regressor = xgb.XGBRegressor(enable_categorical=True, n_estimators=100, max_depth=5, learning_rate=0.1, random_state=42)
xgb_regressor.fit(X_train_scaled, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=True, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=5, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, random_state=42, ...)

Predict on the test set

In [58]:
y_predxgb = xgb_regressor.predict(X_test_scaled)

In [59]:
y_predxgb

array([10331.591, 10265.318, 15647.186, ..., 10115.188,  6169.595,
       25432.598], dtype=float32)

In [60]:
y_test

14668    10000
24614    10000
11096    16000
10424    10000
26007    13000
         ...  
31330     9925
2862      3000
14754    10000
14170     6000
24385    25000
Name: loan_amnt, Length: 6517, dtype: int64

Evaluate the model

In [61]:
mse = mean_squared_error(y_test, y_predxgb)
print(f'The Mean Squared Error of XRegressor Model is {mse}')
#The Mean Squared Error of the XRegressor Model is 164505.5387713405

The Mean Squared Error of XRegressor Model is 168146.45510336247




Another Evaluation

In [62]:
rmse = mse ** 0.5
print(f'The Root Mean Squared Error of the XRegressor => {rmse}')
#The Root Mean Squared Error of the XRegressor => 405.5928238656849

The Root Mean Squared Error of the XRegressor => 410.0566486515765


Save the XGBoost regressor model

In [67]:

joblib.dump(xgb_regressor, 'C:/Users/PC/Documents/sola/others/Telegram Desktop/xgb_regressor_model.pkl')

['C:/Users/PC/Documents/sola/others/Telegram Desktop/xgb_regressor_model.pkl']

# - - - End of XGBoost - - -